core/vm/evmc.go

```diff
@@ -0,0 +1,376 @@
+}
+
+func (host *HostContext) GetStorage(addr common.Address, key common.Hash) common.Hash {
+	env := host.env
+	return env.StateDB.GetState(addr, key)
+}
+
+func (host *HostContext) SetStorage(addr common.Address, key common.Hash, value common.Hash) (status evmc.StorageStatus) {
+	env := host.env
+
+	oldValue := env.StateDB.GetState(addr, key)
+	if oldValue == value {
+		return evmc.StorageUnchanged
+	}
+
+	current := env.StateDB.GetState(addr, key)
+	original := env.StateDB.GetCommittedState(addr, key)
+
+	env.StateDB.SetState(addr, key, value)
+
+	hasNetStorageCostEIP := env.ChainConfig().IsConstantinople(env.BlockNumber) &&
+		!env.ChainConfig().IsPetersburg(env.BlockNumber)
+	if !hasNetStorageCostEIP {
+
+		zero := common.Hash{}
+		status = evmc.StorageModified
+		if oldValue == zero {
+			return evmc.StorageAdded
+		} else if value == zero {
+			env.StateDB.AddRefund(params.SstoreRefundGas)
+			return evmc.StorageDeleted
+		}
+		return evmc.StorageModified
+	}
+
+	if original == current {
+		if original == (common.Hash{}) { // create slot (2.1.1)
+			return evmc.StorageAdded
+		}
+		if value == (common.Hash{}) { // delete slot (2.1.2b)
+			env.StateDB.AddRefund(params.NetSstoreClearRefund)
+			return evmc.StorageDeleted
+		}
+		return evmc.StorageModified
+	}
+	if original != (common.Hash{}) {
+		if current == (common.Hash{}) { // recreate slot (2.2.1.1)
+			env.StateDB.SubRefund(params.NetSstoreClearRefund)
+		} else if value == (common.Hash{}) { // delete slot (2.2.1.2)
+			env.StateDB.AddRefund(params.NetSstoreClearRefund)
+		}
+	}
+	if original == value {
+		if original == (common.Hash{}) { // reset to original inexistent slot (2.2.2.1)
+			env.StateDB.AddRefund(params.NetSstoreResetClearRefund)
+		} else { // reset to original existing slot (2.2.2.2)
+			env.StateDB.AddRefund(params.NetSstoreResetRefund)
+		}
+	}
+	return evmc.StorageModifiedAgain
+}
+
+func (host *HostContext) GetBalance(addr common.Address) common.Hash {
+	env := host.env
+	balance := env.StateDB.GetBalance(addr)
+	return common.BigToHash(balance)
+}
+
+func (host *HostContext) GetCodeSize(addr common.Address) int {
+	env := host.env
+	return env.StateDB.GetCodeSize(addr)
+}
+
+func (host *HostContext) GetCodeHash(addr common.Address) common.Hash {
+	env := host.env
+	if env.StateDB.Empty(addr) {
+		return common.Hash{}
+	}
+	return env.StateDB.GetCodeHash(addr)
+}
+
+func (host *HostContext) GetCode(addr common.Address) []byte {
+	env := host.env
+	return env.StateDB.GetCode(addr)
+}
+
+func (host *HostContext) Selfdestruct(addr common.Address, beneficiary common.Address) {
+	env := host.env
+	db := env.StateDB
+	if !db.HasSuicided(addr) {
+		db.AddRefund(params.SuicideRefundGas)
+	}
+	balance := db.GetBalance(addr)
+	db.AddBalance(beneficiary, balance)
+	db.Suicide(addr)
+}
+
+func (host *HostContext) GetTxContext() (gasPrice common.Hash, origin common.Address, coinbase common.Address,
+	number int64, timestamp int64, gasLimit int64, difficulty common.Hash) {
```

Comment:
Lets define a struct and fill the data. This looks really bad to return 7 values from a method.

Reply:
This will require an EVMC change. Happy to do it, but there might be a hidden issue with returning a struct related to cgo (it was the reason why the `Call()` takes a ton of arguments instead of a single `Message` instance). So I cannot promise it can be done easily.

Reply:
Done in https://github.com/ethereum/evmc/pull/197.

Reply:
Addressed.